src/Modules/LearningTasks/Tutor.LearningTasks.Core/UseCases/Learning/TaskService.cs

```diff
@@ -33,12 +37,24 @@ public Result<LearningTaskDto> Get(int id, int unitId, int learnerId)
         return _mapper.Map<LearningTaskDto>(learningTask);
     }
 
-    public Result<List<LearningTaskDto>> GetByUnit(int unitId, int learnerId)
+    public Result<List<ProgressDto>> GetByUnit(int unitId, int learnerId)
     {
         if (!_accessServices.IsEnrolledInUnit(unitId, learnerId))
             return Result.Fail(FailureCode.Forbidden);
 
-        var learningTasks = _taskRepository.GetNonTemplateByUnit(unitId);
-        return learningTasks.Select(_mapper.Map<LearningTaskDto>).ToList();
+        var tasks = _taskRepository.GetNonTemplateByUnit(unitId);
+        var result = new List<ProgressDto>();
+        foreach (var task in tasks)
+        {
+            ProgressDto progressDto = new ProgressDto { Id = task.Id, Order = task.Order, Name = task.Name!,
+                Status = string.Empty, TotalSteps = task.Steps!.Count(s => s.ParentId == 0) };
+            TaskProgress? taskProgress = _progressRepository.GetByTask(task.Id, learnerId);
```

Comment:
Stop pulling data from the database inside loops.

Create a new method GetByTasks and send an array/list of IDs.